src/pathfinder/opf/opf_ship.cpp

```diff
@@ -195,20 +200,22 @@ Track OPFShipChooseTrack(const Ship *v, TileIndex tile, DiagDirection enterdir,
 	Track track;
 
 	/* Let's find out how far it would be if we would reverse first */
-	Trackdir trackdir = v->GetVehicleTrackdir();
-	TrackBits b = TrackStatusToTrackBits(GetTileTrackStatus(tile2, TRANSPORT_WATER, 0)) & DiagdirReachesTracks(ReverseDiagDir(enterdir)) & TrackdirBitsToTrackBits(TrackdirToTrackdirBits(trackdir));
-
-	uint distr = UINT_MAX; // distance if we reversed
-	if (b != 0) {
-		distr = FindShipTrack(v, tile2, ReverseDiagDir(enterdir), b, tile, &track);
-		if (distr != UINT_MAX) distr++; // penalty for reversing
+	uint rev_dist = UINT_MAX; // distance if we reverse
+	Track cur_track = TrackdirToTrack(v->GetVehicleTrackdir()); // track on the current tile
+	DiagDirection rev_enterdir = ReverseDiagDir(enterdir);
+	TrackBits rev_tracks = TrackStatusToTrackBits(GetTileTrackStatus(tile2, TRANSPORT_WATER, 0)) &
+			DiagdirReachesTracks(rev_enterdir);
+
+	if ((rev_tracks & TrackToTrackBits(cur_track) != TRACK_BIT_NONE) {
+		rev_dist = FindShipTrack(v, tile2, rev_enterdir, TrackToTrackBits(cur_track), tile, &track);
```

Comment:
While the changes are now much harder to follow, this change still changed behaviour.

Previously FindShipTrack was called with "b = GetTileTrackStatus & rev_enterdir & trackdir".
Now it is called with "b = trackdir".

This change now essentially ignores the result of GetTileTrackStatus when calling FindShipTrack.

Reply:
In the original code, FindShipTrack is only called when b contains only the trackbit corresponding to the vehicle trackdir. But the original code is difficult to read, hiding in a single long line with lots of operations that:
- it calculates the trackbits if we enter the tile from a reversed direction (the rev_tracks)
- then it checks that the current vehicle trackbit is on them
- if the rev_tracks contain the trackbit of the vehicle, then it calls FindShipTrack whit an obscure variable (b), whose meaning is not clear (but which is just the current vehicle trackdir).

Yes, the first version of the PR was absolutely wrong, but I fail to see the mistake of the updated PR.

Apart that I may be wrong and I may be changing the behavior of the function, I think that the meaning and use of each variable is easier to understand with the proposed changes.